File: Niveau 3/05 – Tableaux avancés/2 - Carré magique.py
def main():
    n = int(input())
    grid = [0] * n
    b = True
    nbMagic = 0
    for i in range(n): #Remplis la grille, en profite pour calculé le "nombre magique" et vérifié chaque row
        grid[i] = list(map(int, input().split(" ")))
        if nbMagic == 0: nbMagic = sum(grid[i])
        elif nbMagic != sum(grid[i]): b = False
    for col in range(n): #Vérifie chaque colone
        s = 0
        for row in range(n): s += grid[row][col]
        if s != nbMagic: b = False
    sDiag1 = 0
    sDiag2 = 0
    for i in range(n): #Vérifie les diagonales
        sDiag1 += grid[i][i]
        sDiag2 += grid[i][n-i-1]
    if not(sDiag1 == sDiag2 == nbMagic):
        b = False
    for nb in range(n ** 2): #Vérifie si tout les nombres de 1 a n ** 2 se trouve dans la grille
      nbInGrid = False
      for i in grid:
        if nb+1 in i: nbInGrid = True
      if not(nbInGrid): b = False
    if b: print('yes')
    else: print('no')
main()

File: Niveau 3/05 – Tableaux avancés/2 - Carré magique.cpp
#include <iostream>
using namespace std;

static const int MAX_SIZE = 20;

int main()
{
  ios::sync_with_stdio(false);
  int size, nombreMagique = 0;
  bool verifDouble[MAX_SIZE * MAX_SIZE + 1] = {false};
  cin >> size;
  int verifCol[MAX_SIZE] = {0};
  int verifLig[MAX_SIZE] = {0};
  int verifDiag1 = 0;
  int verifDiag2 = 0;
  for (int i = 0; i < size; i++) // Fill la grid, calculé le nombre magique, vérifié les doublons et remplir le tableau pour vérifié les colonnes / lignes
  {
    for (int j = 0; j < size; j++)
    {
      int n;
      cin >> n;
      if (n <= 0 || n > size * size) // Vérifié si le nombre se trouve dans l'intervale 1 --> size * size
      {
        cout << "no";
        return 0;
      }
      if (i == 0) // Calculé le nombre magique si i == 0
      {
        nombreMagique += n;
      }
      if (verifDouble[n - 1]) // Si doublon stopper l'algorithme
      {
        cout << "no";
        return 0;
      }
      if (i == j)
      {
        verifDiag1 += n; // Additionner le nombre a verifDiag1 si il fait partie de la diag1
      }
      if (i + j == size - 1)
      {
        verifDiag2 += n; // Additionner le nombre a verifDiag2 si il fait partie de la diag2
      }
      verifDouble[n - 1] = true; // Ajouter le nombre a la liste qui verifie les doublons
      verifCol[j] += n;          // Additionner le nombre a verifCol[j]
      verifLig[i] += n;          // Additionner le nombre a verifLig[i]
    }
  }

  if (verifDiag1 != nombreMagique || verifDiag2 != nombreMagique) // Verifié les diagonale
  {
    cout << "no";
    return 0;
  }

  for (int i = 0; i < size; i++) // Verifié les lignes et colonnes
  {
    int ligne = verifLig[i];
    int col = verifCol[i];
    if (ligne != nombreMagique || col != nombreMagique)
    {
      cout << "no";
      return 0;
    }
  }

  cout << "yes";
}
